File: driver/src/bootstrap.rs
//! Seeding `batch_commit`'s `batch_id = 0` row: the reset state (`pc =
//! RAM_BASE`, `x1..x31 = 0`), so the first real batch's lookup of the
//! previous batch has a row to read.
//!
//! A fixed-literal insert with no computation, run once before the driver's
//! first batch. Re-running it before any real batch has committed is
//! harmless, since `batch_id = 0` stays the reset state either way, but this
//! is not a recovery step: it writes exactly one row, once, per fresh run.

use clickdoom_spec::RAM_BASE;
use clickhouse::Row;
use serde::Serialize;

use crate::client::{Db, Error};

/// `x1..x31`, SPEC's reset vector.
pub const RESET_REGS: [u32; 31] = [0; 31];

/// One `batch_commit` row's full shape: bootstrap's reset-state seed and a
/// benchmark's snapshot seed both write exactly this.
#[derive(Row, Serialize)]
pub(crate) struct BatchCommitRow {
    pub batch_id: u64,
    pub icount: u64,
    pub pc: u32,
    pub regs: Vec<u32>,
    pub halted: u8,
    pub halt_reason: String,
    pub exit_code: u32,
    pub keyq_pos: u64,
    pub has_frame: u8,
    pub frame_no: u32,
    pub wl_addr: Vec<u32>,
    pub wl_val: Vec<u32>,
    pub wl_icount: Vec<u64>,
    pub fb_wl_addr: Vec<u32>,
    pub fb_wl_val: Vec<u32>,
    pub fb_wl_icount: Vec<u64>,
    pub pal_wl_addr: Vec<u32>,
    pub pal_wl_val: Vec<u32>,
    pub pal_wl_icount: Vec<u64>,
    pub console_bytes: Vec<u8>,
}

#[derive(Debug, thiserror::Error)]
pub enum SeedError {
    #[error("regs holds {0} registers, x1..x31 needs exactly 31")]
    WrongRegisterCount(usize),
    #[error(transparent)]
    Db(#[from] Error),
}

/// What seeding did, for the caller to report.
#[derive(Debug)]
pub enum Seeded {
    /// A `batch_id = 0` row was written, with this many registers.
    Fresh { registers: usize },
    /// `batch_id = 0` already existed; nothing was written.
    AlreadySeeded,
}

/// Seeds `batch_id = 0` with `regs` (`x1..x31`), unless a `batch_id = 0` row
/// already exists.
pub async fn seed(db: &Db, regs: &[u32]) -> Result<Seeded, SeedError> {
    if regs.len() != 31 {
        return Err(SeedError::WrongRegisterCount(regs.len()));
    }

    let existing: u64 = db
        .fetch_one("SELECT count() FROM batch_commit WHERE batch_id = 0")
        .await?;
    if existing > 0 {
        return Ok(Seeded::AlreadySeeded);
    }

    let row = BatchCommitRow {
        batch_id: 0,
        icount: 0,
        pc: RAM_BASE,
        regs: regs.to_vec(),
        halted: 0,
        halt_reason: String::new(),
        exit_code: 0,
        keyq_pos: 0,
        has_frame: 0,
        frame_no: 0,
        wl_addr: Vec::new(),
        wl_val: Vec::new(),
        wl_icount: Vec::new(),
        fb_wl_addr: Vec::new(),
        fb_wl_val: Vec::new(),
        fb_wl_icount: Vec::new(),
        pal_wl_addr: Vec::new(),
        pal_wl_val: Vec::new(),
        pal_wl_icount: Vec::new(),
        console_bytes: Vec::new(),
    };
    db.insert_all("batch_commit", std::iter::once(row)).await?;

    Ok(Seeded::Fresh {
        registers: regs.len(),
    })
}

File: executor/bootstrap.py
#!/usr/bin/env python3
"""Seed `batch_commit`'s batch_id=0 row -- SPEC §1's reset state (`pc =
0x8000_0000`, `x1..x31 = 0`), so the first real batch's `PREV` lookup
(`fold.py`'s `batch()`, `ORDER BY batch_id DESC LIMIT 1`) has a row to read.
A fixed-literal insert with no computation -- PURITY.md action 4
("housekeeping that computes nothing"), same category as the ROM loader
loading ROM bytes: this establishes batch-execution state (SPEC §6), not
ROM/memory state (SPEC §4), so it's its own script rather than folded into
`clickdoom load-rom`.

Run once, before the driver's first batch. Idempotent in the sense that
re-running it is harmless *before* any real batch has committed (batch_id=0
stays the reset state either way); it is NOT meant to be re-run after
batch_id=0 has been superseded -- there is nothing to recover here, unlike
`commit.py`'s flushes, since this writes exactly one row, once, ever, per
fresh run.

Usage:
    bootstrap.py --host localhost --port 9000 --password clickdoom
"""
import argparse
import subprocess
import sys

import config


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--host", default="localhost")
    ap.add_argument("--port", default="9000")
    ap.add_argument("--user", default="default")
    ap.add_argument("--password", default="")
    ap.add_argument("--database", default="clickdoom")
    ap.add_argument("--client", default="clickhouse-client",
                     help="client binary; pass 'clickhouse client' (two words, quoted) "
                          "for the standalone build")
    ap.add_argument("--regs", type=int, nargs=31, default=[0] * 31,
                     help="override the 31-element x1..x31 reset vector (default: all zero, SPEC §1)")
    args = ap.parse_args()

    base_cmd = args.client.split() + [
        "--host", args.host,
        "--port", str(args.port),
        "--user", args.user,
        "--database", args.database,
    ]
    if args.password:
        base_cmd += ["--password", args.password]

    existing = subprocess.run(
        base_cmd + ["--query", "SELECT count() FROM batch_commit WHERE batch_id = 0"],
        capture_output=True, text=True,
    )
    if existing.returncode != 0:
        print(existing.stderr, file=sys.stderr, end="")
        return existing.returncode
    if int(existing.stdout.strip() or "0") > 0:
        print("batch_commit already has a batch_id=0 row -- not seeding again "
              "(this is a fresh-run bootstrap, not a recovery step)", file=sys.stderr)
        return 0

    regs = "[" + ",".join(str(r) for r in args.regs) + "]"
    insert = base_cmd + [
        "--query",
        "INSERT INTO batch_commit "
        "(batch_id, icount, pc, regs, halted, halt_reason, exit_code, "
        " keyq_pos, has_frame, frame_no, wl_addr, wl_val, wl_icount, console_bytes) "
        f"VALUES (0, 0, {config.RAM_BASE}, {regs}, 0, '', 0, "
        f" 0, 0, 0, [], [], [], [])",
    ]
    result = subprocess.run(insert, text=True)
    if result.returncode != 0:
        return result.returncode

    print(f"seeded batch_commit batch_id=0: pc={config.RAM_BASE:#x}, "
          f"{len(args.regs)} registers, icount=0", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
